### PR TITLE
Implement Core attribute Order.

### PR DESCRIPTION
**Pull request by dkarpele**
Implement a possibility to set order of the tests' execution.
The default `order` for an object will be 0, and it's the highest order.
The `order` will be used in `ls`, `show`, `discover` and other methods that will use
`order`.

### CODE REVIEW

**Comment by dkarpele** [comment]
Hi @sopos and @psss,
I've created a draft commit for the issue. The patch is really small, just point me please, whether I'm looking at the right direction.

File: tests/core/order/test.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        for tmt in 'tests' 'plans' 'stories'; do
+          for ((i = 0; i < 5; i++)); do
+            rlRun "tmt $tmt show \"${order[i]}\" | grep order | tee $output"
+            rlAssertGrep "${order_value[i]}" $output
+          done
+        done
+    rlPhaseEnd
+
+    rlPhaseStartTest "Tests order with tmt run discover plan"
+        rlRun "tmt run -v discover plan -n ${order[3]} -q | grep test | tee $output"
```
Comment (psss) [inline tests/core/order/test.sh:39]:
```suggestion
        rlRun "tmt run -q discover -v plan -n ${order[3]} | grep test | tee $output"
```

I'm not sure what `plan --quite` should do, I'm afraid this is not defined and might change :) Let's rather use what's for sure :)

File: tmt/base.py [hunk 2]
```diff
@@ -86,6 +86,12 @@ def __init__(self, node, parent=None):
         # Check whether the node is enabled, handle the default
         self._check('enabled', expected=bool, default=True)
 
+        # Check whether the order is integer, handle the default
+        self._check('order', expected=int, default=0)
+        if self.order < 0:
+            raise tmt.utils.SpecificationError(
+                f"The 'order' attribute must be greater than or equal to zero")
```
Comment (psss) [inline tmt/base.py:93]:
```suggestion
                f"The 'order' attribute must be greater than or equal to zero.")
```

File: tmt/base.py [hunk 3]
```diff
@@ -785,6 +792,7 @@ def show(self):
             echo(tmt.utils.format(
                 'environment', self.environment, key_color='blue'))
         echo(tmt.utils.format('enabled', self.enabled, key_color='cyan'))
+        echo(tmt.utils.format('order', self.order, key_color='cyan'))
```
Comment (psss) [inline tmt/base.py:795]: Perhaps we could show the order only if it is not default?

File: tmt/base.py [hunk 4]
```diff
@@ -86,6 +86,12 @@ def __init__(self, node, parent=None):
         # Check whether the node is enabled, handle the default
         self._check('enabled', expected=bool, default=True)
 
+        # Check whether the order is integer, handle the default
+        self._check('order', expected=int, default=0)
```
Comment (psss) [inline tmt/base.py:90]: Let's create a new constant for the default order and use it here.

```suggestion
        self._check('order', expected=int, default=DEFAULT_ORDER)
```

File: tmt/base.py [hunk 5]
```diff
@@ -1191,6 +1200,7 @@ def _filters_conditions(self, nodes, filters, conditions, links):
                 # Handle broken link as not matching
                 continue
             result.append(node)
+        result = sorted(result, key=lambda n: n.order)
```
Comment (psss) [inline tmt/base.py:1203]: This works fine for most cases, but there's one special case when tests are duplicated:

```yaml
/plan:
    discover:
        how: fmf
        test:
          - test
          - test
    execute:
        how: tmt

/test:
    test: /bin/true

    /one:
        order: 1
    /two:
        order: 2
    /three:
        order: 3
```

This results in the following test discovery:

```
> tmt run discover -v
/var/tmp/tmt/run-056

/plan
    discover
        how: fmf
        directory: /tmp/dup
        tests: test and test
        summary: 6 tests selected
            /test/one
            /test/one
            /test/two
            /test/two
            /test/three
            /test/three
```

We need the following order:

```
/test/one
/test/two
/test/three
/test/one
/test/two
/test/three
```

**Review by psss — CHANGES_REQUESTED** [review]
Thanks much for implementing this! Overall looks very good. Added a couple of suggestions. The last one is a bit tricky: When tests are duplicated during `discover` we should ensure that the order is preserved in each batch of tests added by the name filter.

File: tmt/base.py (hunk 5 above)
Comment (thrix) [reply tmt/base.py:1203]: if `order` is involved can we print something for `discover`?

```
> tmt run discover -v
/var/tmp/tmt/run-056

/plan
    discover
        how: fmf
        directory: /tmp/dup
        tests: test and test
        summary: 6 tests selected [ordered]
            /test/one
            /test/one
            /test/two
            /test/two
            /test/three
            /test/three
```

Comment (psss) [reply tmt/base.py:1203]: > if `order` is involved can we print something for `discover`?

What do you mean by "if order is involved"? Tests are always somehow ordered. As for now alphabetically.

File: tmt/base.py [hunk 6]
```diff
@@ -86,6 +86,12 @@ def __init__(self, node, parent=None):
         # Check whether the node is enabled, handle the default
         self._check('enabled', expected=bool, default=True)
 
+        # Check whether the order is integer, handle the default
+        self._check('order', expected=int, default=0)
+        if self.order < 0:
```
Comment (psss) [inline tmt/base.py:91]: We've discussed this on today's hacking session and the conclusion was that negative numbers should be allowed as well to provide enough space if the number of objects is large.

File: tmt/base.py (hunk 4 above)
Comment (psss) [reply tmt/base.py:90]: Outcome from today's hacking discussion: The default order should be `50` to make it consistent with what we already have defined for multiple `prepare` step configs.

**Comment by dkarpele** [comment]
Thanks for the review @psss and @thrix! Sorry for skipping tuesday's meeting. All changes should be addressed.

**Review by psss — APPROVED** [review]
Thanks much for addressing the issues. Looks good. Added a couple of adjustments in 8d6e38a.

**Pull request merged**